`ProcessTest` fails on macOS-12

10/12 ProcessTests fails due to timeouts when executed on GitHub Actions macOS-12 env. 
https://github.com/scala-native/scala-native/runs/7658264104?check_suite_focus=true

Wojciech, thank you for the Issue and, especially, the log file.  I am going to copy
it below, because I think it will be essential to chasing this.

I believe there are two defects at play here. I believe that `ProcessTest` is
doing __exactly__ what it was designed to do here: yell if there is a problem.
The problem in the base note is, I believe & running from memory, a problem
with Process exit.  I will have to trace, but I believe two of those test
involve explicit signal delivery.

The second defect and one which pops in and out is the "kill signal is
not handled on SN M1".  There is an `AssumptionViolation` in the
SN Repository ProcessTest. I saw nothing in the PR which would
lead me to believe that the ProcessTest in the PR-under-test does
not have that code.  

A big part of a solution here is naming & numbering the beasts.

As a datapoint, given the ProcessTest being reported by the 
Principal & Vice-Principal, I ran Process test on my macOS 12.5 M1.
The `AssumptionViolation`s  triggered, as expected, but not
the mess in the logfile.

Unless someone else reports _this_ failure, we are probably
looking at something in the CI environment.  Oh joy! Meanwhile
saying __soto_voce__ "Thank Fate that is not in my environment!
Don't let that thing come here"

Sorry, ugly logfile extract. Primary data, need to keep it from disappearing
on me.
```
2022-08-03T18:35:04.2449170Z [0m[[0m[0minfo[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[36mshellFallback[0m started[0m
2022-08-03T18:35:04.2573050Z [0m[[0m[0minfo[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[36mdestroyForcibly[0m started[0m
2022-08-03T18:35:04.2674940Z [0m[[0m[31merror[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[31mdestroyForcibly[0m failed: java.lang.[31mAssertionError[0m: process should have exited but timed out (limit: 502 milliseconds), took 0.589 sec[0m
2022-08-03T18:35:04.2778660Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$.$anonfun$currentStackTrace$1(Unknown Source)[0m
2022-08-03T18:35:04.2880270Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$$$Lambda$2.apply(Unknown Source)[0m
2022-08-03T18:35:04.2981830Z [0m[[0m[31merror[0m] [0m[0m    at scala.scalanative.unsafe.Zone$.apply(Unknown Source)[0m
2022-08-03T18:35:04.3083570Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$.currentStackTrace(Unknown Source)[0m
2022-08-03T18:35:04.3185790Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.Throwable.fillInStackTrace(Unknown Source)[0m
2022-08-03T18:35:04.3287340Z [0m[[0m[31merror[0m] [0m[0m    ...[0m
2022-08-03T18:35:04.3387950Z [0m[[0m[0minfo[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[36mdestroy[0m started[0m
2022-08-03T18:35:04.3444160Z [0m[[0m[31merror[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[31mdestroy[0m failed: java.lang.[31mAssertionError[0m: process should have exited but timed out (limit: 501 milliseconds), took 0.516 sec[0m
2022-08-03T18:35:04.3447630Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$.$anonfun$currentStackTrace$1(Unknown Source)[0m
2022-08-03T18:35:04.3451000Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$$$Lambda$2.apply(Unknown Source)[0m
2022-08-03T18:35:04.3453030Z [0m[[0m[31merror[0m] [0m[0m    at scala.scalanative.unsafe.Zone$.apply(Unknown Source)[0m
2022-08-03T18:35:04.3454660Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$.currentStackTrace(Unknown Source)[0m
2022-08-03T18:35:04.3456350Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.Throwable.fillInStackTrace(Unknown Source)[0m
2022-08-03T18:35:04.3457610Z [0m[[0m[31merror[0m] [0m[0m    ...[0m
2022-08-03T18:35:04.3459040Z [0m[[0m[0minfo[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[36mwaitForWithTimeoutTimesOut[0m started[0m
2022-08-03T18:35:04.3460600Z [0m[[0m[0minfo[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[36mwaitForWithTimeoutCompletes[0m started[0m
2022-08-03T18:35:04.3462470Z [0m[[0m[31merror[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[31mwaitForWithTimeoutCompletes[0m failed: java.lang.[31mAssertionError[0m: process should have exited but timed out (limit: 30 seconds), took 30.005 sec[0m
2022-08-03T18:35:04.3464390Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$.$anonfun$currentStackTrace$1(Unknown Source)[0m
2022-08-03T18:35:04.3465770Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$$$Lambda$2.apply(Unknown Source)[0m
2022-08-03T18:35:04.3467130Z [0m[[0m[31merror[0m] [0m[0m    at scala.scalanative.unsafe.Zone$.apply(Unknown Source)[0m
2022-08-03T18:35:04.3469250Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$.currentStackTrace(Unknown Source)[0m
2022-08-03T18:35:04.3470820Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.Throwable.fillInStackTrace(Unknown Source)[0m
2022-08-03T18:35:04.3472160Z [0m[[0m[31merror[0m] [0m[0m    ...[0m
2022-08-03T18:35:04.3473460Z [0m[[0m[0minfo[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[36mredirectErrorStream[0m started[0m
2022-08-03T18:35:04.3475040Z [0m[[0m[31merror[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[31mredirectErrorStream[0m failed: java.lang.[31mAssertionError[0m: Process took more than 10 SECONDS to exit., took 10.134 sec[0m
2022-08-03T18:35:04.3476940Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$.$anonfun$currentStackTrace$1(Unknown Source)[0m
2022-08-03T18:35:04.3478800Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$$$Lambda$2.apply(Unknown Source)[0m
2022-08-03T18:35:04.3480200Z [0m[[0m[31merror[0m] [0m[0m    at scala.scalanative.unsafe.Zone$.apply(Unknown Source)[0m
2022-08-03T18:35:04.3482150Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$.currentStackTrace(Unknown Source)[0m
2022-08-03T18:35:04.3483770Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.Throwable.fillInStackTrace(Unknown Source)[0m
2022-08-03T18:35:04.3485540Z [0m[[0m[31merror[0m] [0m[0m    ...[0m
2022-08-03T18:35:04.3487310Z [0m[[0m[0minfo[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[36moutputStreamReadsFromFile[0m started[0m
2022-08-03T18:35:04.3488830Z [0m[[0m[31merror[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[31moutputStreamReadsFromFile[0m failed: java.lang.[31mAssertionError[0m: Process took more than 10 SECONDS to exit., took 10.008 sec[0m
2022-08-03T18:35:04.3490410Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$.$anonfun$currentStackTrace$1(Unknown Source)[0m
2022-08-03T18:35:04.3491850Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$$$Lambda$2.apply(Unknown Source)[0m
2022-08-03T18:35:04.3493300Z [0m[[0m[31merror[0m] [0m[0m    at scala.scalanative.unsafe.Zone$.apply(Unknown Source)[0m
2022-08-03T18:35:04.3494870Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$.currentStackTrace(Unknown Source)[0m
2022-08-03T18:35:04.3496350Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.Throwable.fillInStackTrace(Unknown Source)[0m
2022-08-03T18:35:04.3497700Z [0m[[0m[31merror[0m] [0m[0m    ...[0m
2022-08-03T18:35:04.3499220Z [0m[[0m[0minfo[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[36minputStreamWritesToFile[0m started[0m
2022-08-03T18:35:04.3500800Z [0m[[0m[31merror[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[31minputStreamWritesToFile[0m failed: java.lang.[31mAssertionError[0m: Process took more than 10 SECONDS to exit., took 10.080 sec[0m
2022-08-03T18:35:04.3502760Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$.$anonfun$currentStackTrace$1(Unknown Source)[0m
2022-08-03T18:35:04.3504670Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$$$Lambda$2.apply(Unknown Source)[0m
2022-08-03T18:35:04.3506240Z [0m[[0m[31merror[0m] [0m[0m    at scala.scalanative.unsafe.Zone$.apply(Unknown Source)[0m
2022-08-03T18:35:04.3507940Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$.currentStackTrace(Unknown Source)[0m
2022-08-03T18:35:04.3509600Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.Throwable.fillInStackTrace(Unknown Source)[0m
2022-08-03T18:35:04.3511210Z [0m[[0m[31merror[0m] [0m[0m    ...[0m
2022-08-03T18:35:04.3512970Z [0m[[0m[0minfo[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[36minputAndErrorStream[0m started[0m
2022-08-03T18:35:04.3514720Z [0m[[0m[31merror[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[31minputAndErrorStream[0m failed: java.lang.[31mAssertionError[0m: Process took more than 10 SECONDS to exit., took 10.069 sec[0m
2022-08-03T18:35:04.3517160Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$.$anonfun$currentStackTrace$1(Unknown Source)[0m
2022-08-03T18:35:04.3519020Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$$$Lambda$2.apply(Unknown Source)[0m
2022-08-03T18:35:04.3520700Z [0m[[0m[31merror[0m] [0m[0m    at scala.scalanative.unsafe.Zone$.apply(Unknown Source)[0m
2022-08-03T18:35:04.3522510Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$.currentStackTrace(Unknown Source)[0m
2022-08-03T18:35:04.3524110Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.Throwable.fillInStackTrace(Unknown Source)[0m
2022-08-03T18:35:04.3525700Z [0m[[0m[31merror[0m] [0m[0m    ...[0m
2022-08-03T18:35:04.3527170Z [0m[[0m[0minfo[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[36mpathPrefixOverride[0m started[0m
2022-08-03T18:35:04.3529510Z [0m[[0m[31merror[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[31mpathPrefixOverride[0m failed: java.lang.[31mAssertionError[0m: Process took more than 10 SECONDS to exit., took 10.139 sec[0m
2022-08-03T18:35:04.3531470Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$.$anonfun$currentStackTrace$1(Unknown Source)[0m
2022-08-03T18:35:04.3533170Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$$$Lambda$2.apply(Unknown Source)[0m
2022-08-03T18:35:04.3534800Z [0m[[0m[31merror[0m] [0m[0m    at scala.scalanative.unsafe.Zone$.apply(Unknown Source)[0m
2022-08-03T18:35:04.3536630Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$.currentStackTrace(Unknown Source)[0m
2022-08-03T18:35:04.3538510Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.Throwable.fillInStackTrace(Unknown Source)[0m
2022-08-03T18:35:04.3540020Z [0m[[0m[31merror[0m] [0m[0m    ...[0m
2022-08-03T18:35:04.3541710Z [0m[[0m[0minfo[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[36mpathOverride[0m started[0m
2022-08-03T18:35:04.3543240Z [0m[[0m[31merror[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[31mpathOverride[0m failed: java.lang.[31mAssertionError[0m: Process took more than 10 SECONDS to exit., took 10.115 sec[0m
2022-08-03T18:35:04.3544990Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$.$anonfun$currentStackTrace$1(Unknown Source)[0m
2022-08-03T18:35:04.3546570Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$$$Lambda$2.apply(Unknown Source)[0m
2022-08-03T18:35:04.3548090Z [0m[[0m[31merror[0m] [0m[0m    at scala.scalanative.unsafe.Zone$.apply(Unknown Source)[0m
2022-08-03T18:35:04.3551270Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$.currentStackTrace(Unknown Source)[0m
2022-08-03T18:35:04.3552800Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.Throwable.fillInStackTrace(Unknown Source)[0m
2022-08-03T18:35:04.3554770Z [0m[[0m[31merror[0m] [0m[0m    ...[0m
2022-08-03T18:35:04.3555720Z [0m[[0m[0minfo[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[36mls[0m started[0m
2022-08-03T18:35:04.3557370Z [0m[[0m[31merror[0m] [0m[0mTest javalib.lang.[33mProcessTest[0m.[31mls[0m failed: java.lang.[31mAssertionError[0m: Process took more than 10 SECONDS to exit., took 10.103 sec[0m
2022-08-03T18:35:04.3558880Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$.$anonfun$currentStackTrace$1(Unknown Source)[0m
2022-08-03T18:35:04.3560290Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$$$Lambda$2.apply(Unknown Source)[0m
2022-08-03T18:35:04.3561570Z [0m[[0m[31merror[0m] [0m[0m    at scala.scalanative.unsafe.Zone$.apply(Unknown Source)[0m
2022-08-03T18:35:04.3563020Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.StackTrace$.currentStackTrace(Unknown Source)[0m
2022-08-03T18:35:04.3564370Z [0m[[0m[31merror[0m] [0m[0m    at java.lang.Throwable.fillInStackTrace(Unknown Source)[0m
2022-08-03T18:35:04.3565480Z [0m[[0m[31merror[0m] [0m[0m    ...[0m
2022-08-03T18:35:04.3566500Z [0m[[0m[0minfo[0m] [0m[0m[34mTest run finished: [0m[31m10 failed[0m[34m, [0m[34m0 ignored[0m[34m, 12 total, 112.434s[0m[0m
2022-08-03T18:35:04.3782910Z [0m[[0m[31merror[0m] [0m[0mFailed: Total 5641, Failed 10, Errors 0, Passed 5579, Skipped 52[0m
2022-08-03T18:35:04.3803420Z [0m[[0m[31merror[0m] [0m[0mFailed tests:[0m
2022-08-03T18:35:04.3804590Z [0m[[0m[31merror[0m] [0m[0m	javalib.lang.ProcessTest[0m
2022-08-03T18:35:10.1140420Z [0m[[0m[31merror[0m] [0m[0m(tests2_12 / Test / [31mtest[0m) sbt.TestsFailedException: Tests unsuccessful[0m
2022-08-03T18:35:10.1205080Z [0m[[0m[31merror[0m] [0m[0mTotal time: 287 s (04:47), completed Aug 3, 2022 6:35:10 PM[0m
2022-08-03T18:35:15.5121230Z ##[error]Process completed with exit code 1.
```

--- a/javalib/src/main/scala/java/lang/process/UnixProcess.scala
+++ b/javalib/src/main/scala/java/lang/process/UnixProcess.scala
@@ -179,28 +179,23 @@ object UnixProcess {
         .map(e => s"${e.getKey()}=${e.getValue()}")
     }
 
-    /*
-     * Use vfork rather than fork to avoid copying the parent process memory to the child. It also
-     * ensures that the parent won't try to read or write to the child file descriptors before the
-     * child process has called execve. In an ideal world, we'd use posix_spawn, but it doesn't
-     * support changing the working directory of the child process or closing all of the unused
-     * parent file descriptors. Using posix_spawn would require adding an additional step in which
-     * we spawned a new process that called execve with a helper binary. This may be necessary
-     * eventually to increase portability but, for now, just use vfork, which is suppported on
-     * OSX and Linux (despite warnings about vfork's future, it seems somewhat unlikely that support
-     * will be dropped soon.
-     */
-    unistd.vfork() match {
+    unistd.fork() match {
       case -1 =>
         throw new IOException("Unable to fork process")
+
       case 0 =>
-        /*
-         * It is unsafe to directly run any code in vfork2 on top of the parent's stack without
-         * creating a new stack frame on the child. To fix this, put all of the code that needs
-         * to run on the child before execve inside of a method.
+        /* 2022-08-06 Old unattributed requirement from prior 'vfork'
+         *     implementation. Probably from older FreeBSD documentation.
+         *     Probably unnecessary but harmless in current 'fork()'
+         *     implementation. Someday a braver soul with time on their hands
+         *     can try removing it.
+         *
+         * It is unsafe to directly run any code in vfork2 on top of the
+         * parent's stack without creating a new stack frame on the child.
+         * To fix this, put all of the code that needs to run on the child
+         * before execve inside of a method.
          */
         def invokeChildProcess(): Process = {
-          ProcessMonitor.notifyMonitor()
           if (dir != null) unistd.chdir(toCString(dir.toString))
           setupChildFDS(!infds, builder.redirectInput(), unistd.STDIN_FILENO)
           setupChildFDS(
@@ -228,12 +223,30 @@ object UnixProcess {
               unistd.execve(c"/bin/sh", newArgv, envp)
             }
           }
-          // The spec of vfork requires calling _exit if the child process fails to execve.
+          /* Older, definitely FreeBSD requirement. No such written
+           * requirement with fork(), but it is probably good idea anyway.
+           *   // The spec of vfork requires calling _exit if the child process
+           *   // fails to execve.
+           */
           unistd._exit(1)
           throw new IOException(s"Failed to create process for command: $cmd")
         }
+
         invokeChildProcess()
+
       case pid =>
+        /* Being here, we know that a child process exists, or had existed.
+         * ProcessMonitor needs to know about it. It is _far_ better
+         * to do the notification in this parent.
+         *
+         * Implementations of 'fork' can be very restrictive about what
+         * can run in the child before it calls one of the 'exec*' methods.
+         * 'notifyMonitor' may or may not follow those rules. Even if it
+         * currently does, that could easily change with future maintenance
+         * make it no longer compliant, leading to shrapnel & wasted
+         * developer time.
+         */
+        ProcessMonitor.notifyMonitor()
         Seq(!(outfds + 1), !(errfds + 1), !infds) foreach unistd.close
         new UnixProcess(pid, builder, infds, outfds, errfds)
     }

--- a/javalib/src/main/scala/java/lang/process/UnixProcess.scala
+++ b/javalib/src/main/scala/java/lang/process/UnixProcess.scala
@@ -1,32 +1,19 @@
 package java.lang.process
 
+import java.lang.ProcessBuilder.Redirect
 import java.io.{File, IOException, InputStream, OutputStream}
+import java.io.FileDescriptor
 import java.util.concurrent.TimeUnit
 import java.util.ScalaOps._
+
 import scala.scalanative.unsigned._
 import scala.scalanative.unsafe._
-import scala.scalanative.libc.{errno => err, signal => sig, _}
-import sig._
+import scala.scalanative.libc.{errno => err, signal => sig}
 import err.errno
-import scala.scalanative.posix.{
-  fcntl,
-  pthread,
-  signal,
-  sys,
-  time,
-  unistd,
-  errno => e
-}
+import scala.scalanative.posix.{fcntl, signal, sys, time, unistd, errno => e}
+import signal.{kill, SIGKILL}
 import time._
 import sys.time._
-import e.ETIMEDOUT
-import UnixProcess._
-import java.lang.ProcessBuilder.Redirect
-import pthread._
-import scala.collection.mutable.ArraySeq
-import scala.scalanative.posix.sys.types.{pthread_cond_t, pthread_mutex_t}
-import java.io.FileDescriptor
-import scala.scalanative.posix
 
 private[lang] class UnixProcess private (
     pid: CInt,
@@ -35,10 +22,9 @@ private[lang] class UnixProcess private (
     outfds: Ptr[CInt],
     errfds: Ptr[CInt]
 ) extends GenericProcess {
-  override def destroy(): Unit = posix.signal.kill(pid, SIGTERM)
+  override def destroy(): Unit = kill(pid, sig.SIGTERM)
 
   override def destroyForcibly(): Process = {
-    import posix.signal._
     kill(pid, SIGKILL)
     this
   }
@@ -76,7 +62,10 @@ private[lang] class UnixProcess private (
       case -1 =>
         val ts = stackalloc[timespec]()
         val tv = stackalloc[timeval]()
-        throwOnError(gettimeofday(tv, null), "Failed to set time of day.")
+        UnixProcess.throwOnError(
+          gettimeofday(tv, null),
+          "Failed to set time of day."
+        )
         val nsec = unit.toNanos(timeout) + TimeUnit.MICROSECONDS.toNanos(tv._2)
         val sec = TimeUnit.NANOSECONDS.toSeconds(nsec)
         ts._1 = tv._1 + sec
@@ -91,7 +80,7 @@ private[lang] class UnixProcess private (
       res = f()
       res match {
         case 0   => _exitValue == -1
-        case res => res != ETIMEDOUT
+        case res => res != e.ETIMEDOUT
       }
     }) ()
     res
@@ -184,63 +173,45 @@ object UnixProcess {
         throw new IOException("Unable to fork process")
 
       case 0 =>
-        /* The ">" quoted section below describes an unattributed requirement
-         * from the prior 'vfork' implementation. The requirement probably
-         * comes from older FreeBSD documentation. It is quoted here to
-         * give historical context to "that is strange, why did they do
-         * that???" invokeChildProcess code. Child code after 'fork()' is
-         * usually inline.
-         *
-         * In the current 'fork()' implementation the inner method is now
-         * probably "unnecessary but harmless". Someday a braver soul with
-         * time on their hands can try removing it.
-         *
-         * > It is unsafe to directly run any code in vfork2 on top of the
-         * > parent's stack without creating a new stack frame on the child.
-         * > To fix this, put all of the code that needs to run on the child
-         * > before execve inside of a method.
-         */
-        def invokeChildProcess(): Process = {
-          if (dir != null) unistd.chdir(toCString(dir.toString))
-          setupChildFDS(!infds, builder.redirectInput(), unistd.STDIN_FILENO)
-          setupChildFDS(
-            !(outfds + 1),
-            builder.redirectOutput(),
-            unistd.STDOUT_FILENO
-          )
-          setupChildFDS(
-            !(errfds + 1),
-            if (builder.redirectErrorStream()) Redirect.PIPE
-            else builder.redirectError(),
-            unistd.STDERR_FILENO
-          )
-          unistd.close(!infds)
-          unistd.close(!(infds + 1))
-          unistd.close(!outfds)
-          unistd.close(!(outfds + 1))
-          unistd.close(!errfds)
-          unistd.close(!(errfds + 1))
+        if (dir != null)
+          unistd.chdir(toCString(dir.toString))
 
-          binaries.foreach { b =>
-            val bin = toCString(b)
-            if (unistd.execve(bin, argv, envp) == -1 && errno == e.ENOEXEC) {
-              val newArgv = nullTerminate(Seq("/bin/sh", "-c", b))
-              unistd.execve(c"/bin/sh", newArgv, envp)
-            }
+        setupChildFDS(!infds, builder.redirectInput(), unistd.STDIN_FILENO)
+        setupChildFDS(
+          !(outfds + 1),
+          builder.redirectOutput(),
+          unistd.STDOUT_FILENO
+        )
+        setupChildFDS(
+          !(errfds + 1),
+          if (builder.redirectErrorStream()) Redirect.PIPE
+          else builder.redirectError(),
+          unistd.STDERR_FILENO
+        )
+        unistd.close(!infds)
+        unistd.close(!(infds + 1))
+        unistd.close(!outfds)
+        unistd.close(!(outfds + 1))
+        unistd.close(!errfds)
+        unistd.close(!(errfds + 1))
+
+        binaries.foreach { b =>
+          val bin = toCString(b)
+          if (unistd.execve(bin, argv, envp) == -1 && errno == e.ENOEXEC) {
+            val newArgv = nullTerminate(Seq("/bin/sh", "-c", b))
+            unistd.execve(c"/bin/sh", newArgv, envp)
           }
-
-          /* execve failed. FreeBSD "man" recommends fast exit.
-           * Linux says nada.
-           * Code 127 is "Command not found", the convention for exec failure.
-           */
-          unistd._exit(127)
-          throw new IOException(s"Failed to create process for command: $cmd")
         }
 
-        invokeChildProcess()
+        /* execve failed. FreeBSD "man" recommends fast exit.
+         * Linux says nada.
+         * Code 127 is "Command not found", the convention for exec failure.
+         */
+        unistd._exit(127)
+        throw new IOException(s"Failed to create process for command: $cmd")
 
       case pid =>
-        /* Being here, we know that a child process exists, or had existed.
+        /* Being here, we know that a child process exists, or existed.
          * ProcessMonitor needs to know about it. It is _far_ better
          * to do the notification in this parent.
          *
@@ -258,7 +229,7 @@ object UnixProcess {
   }
 
   @inline
-  private[lang] def throwOnError(rc: CInt, msg: => String): CInt = {
+  private def throwOnError(rc: CInt, msg: => String): CInt = {
     if (rc != 0) {
       throw new IOException(s"$msg Error code: $rc, Error number: $errno")
     } else {

--- a/javalib/src/main/scala/java/lang/process/UnixProcess.scala
+++ b/javalib/src/main/scala/java/lang/process/UnixProcess.scala
@@ -190,7 +190,7 @@ object UnixProcess {
          * give historical context to "that is strange, why did they do
          * that???" invokeChildProcess code. Child code after 'fork()' is
          * usually inline.
-         * 
+         *
          * In the current 'fork()' implementation the inner method is now
          * probably "unnecessary but harmless". Someday a braver soul with
          * time on their hands can try removing it.

--- a/javalib/src/main/scala/java/lang/process/UnixProcess.scala
+++ b/javalib/src/main/scala/java/lang/process/UnixProcess.scala
@@ -184,16 +184,21 @@ object UnixProcess {
         throw new IOException("Unable to fork process")
 
       case 0 =>
-        /* 2022-08-06 Old unattributed requirement from prior 'vfork'
-         *     implementation. Probably from older FreeBSD documentation.
-         *     Probably unnecessary but harmless in current 'fork()'
-         *     implementation. Someday a braver soul with time on their hands
-         *     can try removing it.
+        /* The ">" quoted section below describes an unattributed requirement
+         * from the prior 'vfork' implementation. The requirement probably
+         * comes from older FreeBSD documentation. It is quoted here to
+         * give historical context to "that is strange, why did they do
+         * that???" invokeChildProcess code. Child code after 'fork()' is
+         * usually inline.
+         * 
+         * In the current 'fork()' implementation the inner method is now
+         * probably "unnecessary but harmless". Someday a braver soul with
+         * time on their hands can try removing it.
          *
-         * It is unsafe to directly run any code in vfork2 on top of the
-         * parent's stack without creating a new stack frame on the child.
-         * To fix this, put all of the code that needs to run on the child
-         * before execve inside of a method.
+         * > It is unsafe to directly run any code in vfork2 on top of the
+         * > parent's stack without creating a new stack frame on the child.
+         * > To fix this, put all of the code that needs to run on the child
+         * > before execve inside of a method.
          */
         def invokeChildProcess(): Process = {
           if (dir != null) unistd.chdir(toCString(dir.toString))
@@ -223,12 +228,12 @@ object UnixProcess {
               unistd.execve(c"/bin/sh", newArgv, envp)
             }
           }
-          /* Older, definitely FreeBSD requirement. No such written
-           * requirement with fork(), but it is probably good idea anyway.
-           *   // The spec of vfork requires calling _exit if the child process
-           *   // fails to execve.
+
+          /* execve failed. FreeBSD "man" recommends fast exit.
+           * Linux says nada.
+           * Code 127 is "Command not found", the convention for exec failure.
            */
-          unistd._exit(1)
+          unistd._exit(127)
           throw new IOException(s"Failed to create process for command: $cmd")
         }
 
@@ -314,7 +319,8 @@ object UnixProcess {
     }
   }
 
-  // The execvpe function isn't available on all platforms so find the possible binaries to exec.
+  // The execvpe function isn't available on all platforms so find the
+  // possible binaries to exec.
   private def binaryPaths(
       environment: java.util.Map[String, String],
       bin: String

--- a/nativelib/src/main/resources/scala-native/platform/posix/process_monitor.cpp
+++ b/nativelib/src/main/resources/scala-native/platform/posix/process_monitor.cpp
@@ -57,11 +57,14 @@ static void *wait_loop(void *arg) {
         if (pid != -1) {
             pthread_mutex_lock(&shared_mutex);
 
-            const int last_result = WIFEXITED(status) ? WEXITSTATUS(status)
-                                    : WIFSIGNALED(status)
-                                        ? 0x80 + WTERMSIG(status)
-                                        : status; // other cases probably yield
-                                                  // garbage but let it flow.
+            // probably yields garbage if not superseded below. Let it show.
+            int last_result = status;
+
+            if (WIFEXITED(status)) {
+                last_result = WEXITSTATUS(status);
+            } else if (WIFSIGNALED(status)) {
+                last_result = 0x80 + WTERMSIG(status);
+            }
 
             const auto monitor = waiting_procs.find(pid);
             if (monitor != waiting_procs.end()) {

--- a/nativelib/src/main/resources/scala-native/platform/posix/process_monitor.cpp
+++ b/nativelib/src/main/resources/scala-native/platform/posix/process_monitor.cpp
@@ -56,8 +56,13 @@ static void *wait_loop(void *arg) {
         const int pid = waitpid(-1, &status, 0);
         if (pid != -1) {
             pthread_mutex_lock(&shared_mutex);
-            const int last_result =
-                WIFSIGNALED(status) ? 0x80 + status : status;
+
+            const int last_result = WIFEXITED(status) ? WEXITSTATUS(status)
+                                    : WIFSIGNALED(status)
+                                        ? 0x80 + WTERMSIG(status)
+                                        : status; // other cases probably yield
+                                                  // garbage but let it flow.
+
             const auto monitor = waiting_procs.find(pid);
             if (monitor != waiting_procs.end()) {
                 auto m = monitor->second;

--- a/unit-tests/shared/src/test/scala/javalib/lang/ProcessTest.scala
+++ b/unit-tests/shared/src/test/scala/javalib/lang/ProcessTest.scala
@@ -197,16 +197,64 @@ class ProcessTest {
       proc.destroyForcibly()
   }
 
-  @Test def destroy(): Unit = {
-    assumeFalse(
-      // Fails with traceback on mac arm64 and maybe others.
-      // See Issue #2648
-      "Test is available on arm64 hardware only when using JVM",
-      Platform.hasArm64SignalQuirk
-    )
-    val proc = processSleep(2.0).start()
+  private def processForDestruction(): Process = {
+    /* Return a Process that is suitable to receive a SIGTERM or SIGKILL
+     * signal and return that signal as its exit code.
+     *
+     * The underlying operating system (OS) process must be in the prime of
+     * its life; not too young, not too old.
+     *
+     * Specifically,the signal must be delivered after OS process calls one
+     * of the 'exec' family and before it completes on its own and exits
+     * with an "unexpected" exit code.
+     *
+     * See Issue #2759 for an extended discussion.
+     */
+
+    /* Whenever you see timing code like this, you know both that you
+     * are in for a good time in the present and that you are destined
+     * to return to it time & time again.
+     *
+     * initialDelaySeconds:
+     *   With the current, a hack but done, wait-for-ready implementation
+     *   this value is chosen to drastically decrease the chance of
+     *   the signal arriving before the 'exec' and causing failure.
+     *   Agreed, it is currently wasteful of CI time but the hope
+     *   is that it will not incur developer time chasing race
+     *   issues.
+     *
+     * lifetimeSeconds:
+     *   It is OK for lifetimeSeconds to be on the large side. The
+     *   returned process is destined to be destroyed on receipt.
+     *   Making it too small makes the window for receiving the signal
+     *   smaller, increasing the chance for the signal not hitting the
+     *   window and reporting errors.  Hence, the high side is the better path.
+     *
+     * Again, see Issue #2759 for an extended discussion.
+     */
+
+    val initialDelaySeconds = 5
+    val lifetimeSeconds = 10.0 + initialDelaySeconds
+
+    val proc = processSleep(lifetimeSeconds).start()
 
     assertTrue("process should be alive", proc.isAlive)
+
+    /* Give time for OS child to get past exec*() call.
+     * Agreed, this hasty implementation is sub-optimal & wasteful but
+     * it has the saving virtue of being implemented.
+     *
+     * A better implementation is left as an exercise for the reader.
+     * See Issue #2759 for more complex alternatives.
+     */
+    Thread.sleep(initialDelaySeconds * 1000)
+
+    return proc
+  }
+
+  @Test def destroy(): Unit = {
+    val proc = processForDestruction()
+
     proc.destroy()
 
     val timeout = 501 // Make message distinguished.
@@ -223,22 +271,15 @@ class ProcessTest {
   }
 
   @Test def destroyForcibly(): Unit = {
-    assumeFalse(
-      // Fails with traceback on mac arm64 and maybe others.
-      // See Issue #2648
-      "Test is available on arm64 hardware only when using JVM",
-      Platform.hasArm64SignalQuirk
-    )
-    val proc = processSleep(2.0).start()
+    val proc = processForDestruction()
 
-    assertTrue("process should be alive", proc.isAlive)
-    val p = proc.destroyForcibly()
+    proc.destroyForcibly()
 
     val timeout = 502 // Make message distinguished.
     assertTrue(
       "process should have exited but timed out" +
         s" (limit: ${timeout} milliseconds)",
-      p.waitFor(timeout, TimeUnit.MILLISECONDS)
+      proc.waitFor(timeout, TimeUnit.MILLISECONDS)
     )
     assertEquals(
       // SIGKILL, use unix signal 'excess 128' convention on non-Windows.

